specs/chugsplash-manager.md

```diff
@@ -0,0 +1,76 @@
+
+```typescript
+enum BundleStatus {
+  EMPTY,
+  PROPOSED,
+  APPROVED,
+  CANCELLED,
+  COMPLETED
+}
+```
+
+## Computing a unique bundle ID
+
+```typescript
+const computeBundleId = (
+  bundleRoot: bytes32,
+  bundleSize: uint256,
+  configUri: string
+): bytes => {
+  return keccak256(
+    abi.encode(
+      bundleRoot,
+      bundleSize,
+      configUri,
+    )
+  )
+}
+```
+
+## Proposing a ChugSplash bundle
+
+* A ChugSplash bundle MUST only be proposable by the `admin` of the `ChugSplashManager`.
+* Bundles MUST be in the `EMPTY` state before they can be proposed.
+* Proposing a bundle puts the bundle in the `PROPOSED` state.
+* The ChugSplashManager MUST emit the event ChugSplashBundleProposed with the ID of the bundle, the bundle root, the bundle size, and the config URI as parameters. The ChugSplashManager MUST announce this event to the registry.
+
+## Approving a ChugSplash bundle
+
+* A ChugSplash bundle MUST only be approvable by the `admin` of the `ChugSplashManager`.
+* Bundles MUST have be in the `PROPOSED` state before they can be approved.
+* There MUST NOT be any active bundle.
+* Approving a bundle MUST put the bundle in the `APPROVED` state.
+* Approving a bundle MUST make the approved bundle the active bundle.
+* The balance of the ChugSplashManager minus the current `totalDebt` MUST be greater than `BUNDLE_EXECUTION_BOND_AMOUNT` ETH in its balance.
+* The ChugSplashManager MUST emit the event ChugSplashBundleApproved with the ID of the bundle as the only parameter. The ChugSplashManager MUST announce this event to the registry.
+
+## Cancelling a ChugSplash bundle
+
+* A ChugSplash bundle MUST only be cancellable by the `admin` of the `ChugSplashManager`.
+* Bundles MUST be in the `APPROVED` state before they can be cancelled.
+* Cancelling a bundle MUST put the bundle in the `CANCELLED` state.
+* Cancelling a bundle MUST remove the active bundle.
+* The ChugSplashManager MUST increase the `debt` owed to the current executor by `BUNDLE_EXECUTION_BOND_AMOUNT`.
+* The ChugSplashManager MUST increase the `totalDebt` by the `BUNDLE_EXECUTION_BOND_AMOUNT`.
+* The ChugSplashManager MUST emit the event ChugSplashBundleCancelled with the ID of the bundle and the total number of actions executed. The ChugSplashManager MUST announce this event to the registry.
+
+## Withdrawing ETH
+
+* ETH stored in the contract MUST only be withdrawable by the `admin` of the `ChugSplashManager`.
+* The admin ETH withdrawal function MUST NOT be callable while a bundle is in the `APPROVED` state.
+* The admin ETH withdrawal function MUST only allow the admin to withdraw the balance of the account minus `totalDebt`.
+* The ChugSplashManager MUST emit the ETHWithdrawn event with the recipient and the amount. The ChugSplashManager MUST announce this event to the registry.
+
+## Claiming executor payments
+
+* Executors MUST be able to claim payments at any time.
+* Executors MUST NOT be able to claim more payments than their `debt` amount.
```

Comment:
We might want to change this line to the following:
* If the caller is the `selectedExecutor`, they MUST NOT be able to claim more payments than their `debt` amount minus the `executorBondAmount`.
* Otherwise, the executor MUST NOT be able to claim more payments than their `debt` amount.

Reply:
I think we can get around this if we add `executorBondAmount` to `totalDebt` immediately but only to the executor's `debt` when the execution is completed.

Reply:
Yeah that totally works since you can also just subtract the `executorBondAmount` from the `totalDebt` when a new executor overrides a previous executor instead of doing a payment.

Reply:
> I think we can get around this if we add `executorBondAmount` to `totalDebt` immediately but only to the executor's `debt` when the execution is completed.

Made this change

> Yeah that totally works since you can also just subtract the `executorBondAmount` from the `totalDebt` when a new executor overrides a previous executor instead of doing a payment.

I actually don't think we need to subtract the `executorBondAmount` from the `totalDebt` when a new executor overrides a previous executor. I think we can just leave the `totalDebt` as it currently is. See my comment [here](https://github.com/smartcontracts/chugsplash/pull/60#discussion_r974850462)